source/connection-monitoring-and-pooling/tests/pool-clear-clears-waitqueue.yml

```diff
@@ -0,0 +1,54 @@
+version: 1
+style: unit
+description: clearing pool clears the WaitQueue
+poolOptions:
+  maxPoolSize: 1
+  waitQueueTimeoutMS: 1000
+operations:
+  - name: ready
+  # check out only connection in the pool
+  - name: checkOut
+  # start a few threads that all will enter
+  # the wait queue
+  - name: start
+    target: thread1
+  - name: checkOut
+    thread: thread1
+  - name: start
+    target: thread2
+  - name: checkOut
+    thread: thread2
+  - name: start
+    target: thread3 
+  - name: checkOut
+    thread: thread3
+  # clear the pool, ejecting the previous three threads
```

Comment:
There is a race here where the clear() can happen before the 3 threads emit ConnectionCheckOutStarted which messes up the order in the event assertions. I suggest adding:
```
  # wait for the 3 threads to start checking out a connection
  - name: waitForEvent
    event: ConnectionCheckOutStarted
    count: 4
  # clear the pool, ejecting the previous three threads
  # from the WaitQueue
  - name: clear
  # The 3 threads are immediately ejected
  - name: waitForEvent
    event: ConnectionCheckOutFailed
    count: 4
```

Also, I believe this test will succeed regardless if the driver clears the wait queue or not because the checkout attempts will timeout after 1000ms. To fix this I think we need to check the original connection back in:
```
# ... continuing snippet from above
  - name: checkIn
    connection: conn
```

Reply:
Good catch, fixed.

Note: I put the check in before waiting for the events so that the even expectations can ensure the evictions happen immediately.

Reply:
Hmm I think there's still one more problem. In sync drivers pool.clear() notifies the waiting threads to wake up but doesn't wait for them. This means that the checkIn races with the threads and the order of the ConnectionCheckOutFailed/ConnectionCheckedIn events is non-deterministic. 

I think the only way to fix this is to change the test to use a high waitQueueTimeoutMS (>10seconds) and ignore the ConnectionCheckedIn event.

Reply:
Fixed.

Reply:
Upon reviewing this again I think we need to remove the `checkIn` and increase waitQueueTimeoutMS from 10000 to 30000. Including the `checkIn` hides a bug in the python implementation. The bug is that checkIn wakes up the waiting threads but clear() does not.

Reply:
Okay, `waitForEvent` to include an optional timeout used to ensure the events are seen quickly (within 1s which seemed generous enough).